subsys/bluetooth/audio/has.c

```diff
@@ -0,0 +1,1330 @@
+			       read_active_preset_id, NULL, NULL),
+	BT_GATT_CCC(ccc_cfg_changed, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE_ENCRYPT));
+
+static void process_control_point_tx_work(struct k_work *work);
+
+static struct has_client *client_get(struct bt_conn *conn)
+{
+	struct has_client *client = NULL;
+
+	BT_DBG("conn %p", conn);
+
+	for (int i = 0; i < ARRAY_SIZE(has_client_list); i++) {
+		if (conn == has_client_list[i].conn) {
+			return &has_client_list[i];
+		}
+
+		/* first free slot */
+		if (!client && !has_client_list[i].conn) {
+			client = &has_client_list[i];
+		}
+	}
+
+	if (client) {
+		client->conn = bt_conn_ref(conn);
+		k_work_init_delayable(&client->cp_tx_work, process_control_point_tx_work);
+	}
+
+	BT_DBG("new client %p", client);
+
+	return client;
+}
+
+static inline int preset_changed_popcount(struct has_client *client)
+{
+#if CONFIG_BT_HAS_PRESET_CNT > 32
+	int i, count = 0;
+
+	for (i = 0; i < ARRAY_SIZE(client->preset_changed.pending); i++) {
+		count += popcount(atomic_get(&client->preset_changed.pending[i]));
+	}
+#else
+	return popcount(atomic_get(client->preset_changed.pending));
+#endif
+}
+
+static bool is_preset_changed_pending(struct has_client *client)
+{
+	return preset_changed_popcount(client) > 0;
+}
+
+typedef bool (*preset_func_t)(struct has_preset *preset, void *user_data);
+
+static void preset_foreach(uint8_t start_index, uint8_t end_index, uint8_t num_matches,
+			   preset_func_t func, void *user_data)
+{
+	if (!num_matches) {
+		num_matches = UINT8_MAX;
+	}
+
+	if (start_index <= last_preset_id) {
+		for (int i = 0; i < ARRAY_SIZE(has_preset_list); i++) {
+			struct has_preset *preset = &has_preset_list[i];
+			bool match;
+
+			/* Skip ahead if index is not within requested range */
+			if (preset->id < start_index) {
+				continue;
+			}
+
+			if (preset->id > end_index) {
+				return;
+			}
+
+			match = func(preset, user_data);
+			if (!match) {
+				continue;
+			}
+
+			if (--num_matches > 0) {
+				continue;
+			}
+
+			return;
+		}
+	}
+}
+
+static uint8_t preset_index(struct has_preset *preset)
+{
+	ptrdiff_t index = preset - has_preset_list;
+
+	__ASSERT(index >= 0 && index < ARRAY_SIZE(has_preset_list), "Invalid preset pointer");
+
+	return (uint8_t)index;
+}
+
+static struct has_preset *preset_get(uint8_t index)
+{
+	for (int i = 0; i < ARRAY_SIZE(has_preset_list); i++) {
+		struct has_preset *preset = &has_preset_list[i];
+
+		if (preset->id == index) {
+			return preset;
+		}
+	}
+
+	return NULL;
+}
+
+static void active_preset_work_process(struct k_work *work)
+{
+	bt_gatt_notify_uuid(NULL, BT_UUID_HAS_ACTIVE_PRESET_INDEX, has_svc.attrs,
+			    &has_local.active_id, 1);
+}
+
+static bool is_read_preset_rsp_pending(struct has_client *client)
+{
+	return client->read_preset_rsp.preset_pending != NULL &&
+	       client->read_preset_rsp.num_presets > 0;
+}
+
+static void control_point_tx_work_submit(struct has_client *client, k_timeout_t delay)
```

Comment:
Correct, but you use the same naming scheme for the client, where in the context of the client it refers to sending a control point operation. 

I don't feel using the same name for the server and client, when they are doing two different operation, is a good choice. I am tempted to suggest `control_point_rsp_work_submit` instead (or `cp_rsp_work_submit`), but `rsp` is not really true to the operation either, as it will be a notification or indication, and not really a response (although given the design of HAS, it acts as a response).

Reply:
I renamed it to control_point_write for client (as it uses ATT Write Request) and control_point_send for server (notification/indication)